src/osdc/Objecter.cc

```diff
@@ -3415,36 +3392,53 @@ void Objecter::list_nobjects(NListContext *list_context, Context *onfinish)
   }
   int pg_num = pool->get_pg_num();
   bool sort_bitwise = osdmap->test_flag(CEPH_OSDMAP_SORTBITWISE);
-  rl.unlock();
 
-  if (list_context->starting_pg_num == 0) {     // there can't be zero pgs!
-    list_context->starting_pg_num = pg_num;
+  if (list_context->pos.is_min()) {
+    list_context->starting_pg_num = 0;
     list_context->sort_bitwise = sort_bitwise;
-    ldout(cct, 20) << pg_num << " placement groups" << dendl;
+    list_context->starting_pg_num = pg_num;
   }
   if (list_context->sort_bitwise != sort_bitwise) {
-    ldout(cct, 10) << " hobject sort order changed, restarting this pg"
-		   << dendl;
-    list_context->cookie = collection_list_handle_t();
+    list_context->pos = hobject_t(
```

Comment:
actually, looking at this more closely, the list_nobjects() Objecter interface doesn't have a way to specify a snap and never has.  So CEPH_NOSNAP is correct in this case.  If we wanted to change the API to support listing a snap we could (and IIRC there have been PRs in the past to do this) but until that happens i'm pretty sure NOSNAP is the right thing.  In particular, the arg is "start after", and we give it the last item we got before, so we want it to be == (not <) that previous item.. if we change to 0 here i think we'll get dups on the request boundaries.